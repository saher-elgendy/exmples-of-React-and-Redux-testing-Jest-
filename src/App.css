body {
  background: #faf8f8
}

.App {
  text-align: center;
  background: #faf8f8;
}

img {
  max-width: 100%;
  object-fit: contain;
  cursor: pointer;
}

h4 {
  margin: 0.3rem
}

ul {
  padding: 0;
  margin: 0;
  list-style: none;
}

header {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  padding: 1.5rem;
}

.invisible-sperator {
  visibility: hidden;
}
.bag-container {
  width: 50px;
  object-fit: contain;
}

.book-count {
  width: 10px;
  height: 10px;
  background: red;
  color: #fff;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 6px
}

.books-list {
  width: 100%;
  display: flex;
  flex-wrap: wrap;
}
.book-card {
  padding: 1rem;
  margin: 1rem;
  width: 250px;
  height: 350px;
  display: flex;
  flex-direction: column;
  border: 2px solid #eee;
  margin-top: 1rem;
  background: #fff;
  flex-grow: 1;
  /* box-shadow: 5px 5px 5px 4px; */
}

.book-cover {
  height: 10rem;
  display: block;
  margin: auto;
}

.book-card-btn {
  display: block;
  border: none;
  padding:  1rem;
  font-weight: bold;
  color: #fff;
  background: dodgerblue;
  cursor: pointer;
  bottom: 1rem;
  margin-top: 0.25rem;
  outline: none;
  transition: background 0.2s ease-in-out;
}

.book-card-btn:hover {
  background: #1b64ac
}

.remove-book-btn {
  background: #0b3e6f
}

.bag-content {
  min-width: 200px;
  min-height: 200px;
  max-width: 80%;
  max-height: 350px;
  overflow: auto;
  background: #fff;
  border: 2px solid #000;
  padding: 0.5rem;
  position: absolute;
  margin:auto
}

.bag-item {
  display: flex;
  align-items: center;
  padding: 0.3rem;
  cursor: pointer;
}

.bag-item:hover {
  background: #eee
}

.bag-book-cover {
  width: 30px;
  margin-right: 10px;
}

.book-details {
  text-align: left
}