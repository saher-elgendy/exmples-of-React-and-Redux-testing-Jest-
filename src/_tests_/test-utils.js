import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares);


//creating a custom render function so we can wrap our App with
//react-redux provider
const render = (ui, initialState) => {
  const store = mockStore(initialState);
  //App wrapper and mocked store passed to it
  const Wrapper = ({ children }) => {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }