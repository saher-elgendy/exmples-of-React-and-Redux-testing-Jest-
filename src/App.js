import React, { useEffect } from 'react';
import './App.css';
import { connect } from 'react-redux'
import { fetchData } from './redux/actions';
import Bag from './components/bag';
import Bookstore from './components/bookStore';


function App({ fetchData, data: { loading, error, books } }) {
  useEffect(() => {
    fetchData()
  }, []);

  return (
    <div className="App">
      <header data-testid="header">
        <h2>React Testing</h2>
        <Bag />
      </header>
      {
        error ? error :
          !loading ? <Bookstore books={books} /> : '...loading books'
      }

    </div>
  );
}

const mapStateToProps = state => ({
  data: state.data,
});

export default connect(mapStateToProps, { fetchData })(App);
