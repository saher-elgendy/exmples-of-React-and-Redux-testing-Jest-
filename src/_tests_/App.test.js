import React from 'react';
import App from '../App';
import '@testing-library/jest-dom/extend-expect'
import { screen, render } from './test-utils';


//Testing App Component
describe('App', () => {
  const { getByText, getByTestId, queryByText } = screen;

  let initialState = {
    data: {
      books: [],
      error: '',
      loading: false
    },
    bag: []
  }

  it('should render App correctly with given redux state', () => {
    const { container } = render(<App />, initialState);
    //creatin App component snapshot
    expect(container.firstChild).toMatchSnapshot();
    //test if header exist in the document with the right text
    expect(getByTestId(/header/)).toHaveTextContent('React Testing');
    //when loading is false, remove loading message or don't show it
    expect(queryByText(/...loading books/)).not.toBeInTheDocument()
  });
  //the App should display a loading message before data get fetched
  it('displays loading message before data get fetched', () => {
    initialState = {
      ...initialState,
      data: {
        ...initialState.data,
        loading: true
      }
    }
    render(<App />, initialState);
    expect(queryByText(/...loading books/)).toBeInTheDocument();
  });

  //The App should display an error message if something went wrong while fetching data
  it('display an error message if any thing wrong happened while fetching data', () => {
    initialState = {
      ...initialState,
      data: {
        ...initialState.data,
        error: 'something went wrong'
      }
    }
    render(<App />, initialState);
    expect(getByText(/something went wrong/)).toBeInTheDocument();
  })
})