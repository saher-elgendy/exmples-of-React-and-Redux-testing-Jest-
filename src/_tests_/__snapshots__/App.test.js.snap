// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render App correctly with given redux state 1`] = `
<div
  class="App"
>
  <header
    data-testid="header"
  >
    <h2>
      React Testing
    </h2>
    <div
      class="bag"
    >
      <div
        class="bag-container"
      >
        <span
          class="book-count"
        >
          0
        </span>
        <img
          alt="bag"
          src="icon.png"
        />
      </div>
    </div>
  </header>
  <div
    class="book-store"
  >
    <ul
      class="books-list"
    />
  </div>
</div>
`;
